interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/IdValidator.scala

```diff
@@ -0,0 +1,92 @@
+package pl.touk.nussknacker.engine.compile
+
+import cats.data.Validated.{invalid, valid}
+import cats.data.{NonEmptyList, Validated, ValidatedNel}
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError._
+import pl.touk.nussknacker.engine.canonicalgraph.CanonicalProcess
+import pl.touk.nussknacker.engine.graph.node.NodeData
+
+object IdValidator {
+
+  import cats.implicits._
+
+  private val nodeIdIllegalCharacters = Set('.', '"', '\'')
+
+  def validate(process: CanonicalProcess): ValidatedNel[ProcessCompilationError, Unit] = {
+    val scenarioIdValidationResult =
+      validateIdIsNotEmpty(process).andThen { _ =>
+        validateIdIsNotBlank(process).andThen { _ =>
+          (validateIdHasNoLeadingSpaces(process), validateIdHasNoTrailingSpaces(process)).mapN((_, _) => ())
+        }
+      }
+
+    val nodesIdValidationResult = process.nodes
+      .map(node => validateNodeId(node.data))
+      .foldLeft(Validated.validNel[ProcessCompilationError, Unit](())) { (acc, validation) =>
+        acc.combine(validation)
+      }
+
+    scenarioIdValidationResult.combine(nodesIdValidationResult)
+  }
+
+  def validateNodeId(node: NodeData): ValidatedNel[ProcessCompilationError, Unit] = {
+    (
+      validateIdIsNotEmpty(node).andThen { _ =>
+        validateIdIsNotBlank(node).andThen { _ =>
+          (validateIdHasNoLeadingSpaces(node), validateIdHasNoTrailingSpaces(node)).mapN((_, _) => ())
+        }
+      },
+      validateNodeHasNoIllegalCharacters(node)
+    )
+      .mapN((_, _) => ())
+  }
+
+  private def applySingleErrorValidation[T](
+      predicate: T => Boolean,
+      errorProducer: T => ProcessCompilationError
+  )(implicit process: T): ValidatedNel[ProcessCompilationError, T] = {
+    if (predicate(process))
+      invalid(NonEmptyList.one(errorProducer(process)))
+    else
+      valid(process)
+  }
+
+  private def validateIdIsNotEmpty(implicit process: CanonicalProcess) =
```

Comment:
Is there some reason why you do not just pass id to to these methods instead of extracting it in each method? Regarding fragment/scenario distinguishing in errors, we can pass some implicit parameter for that purpose.

Reply:
No, there's no reason besides differentiating node/process and fragment/scenario errors. Now I've created a private class for scenario id validation `ScenarioIdValidationData(scenarioId: String, isFragment: Boolean)` not to pass entire CanonicalProcess. 

Or did you mean that the code would be cleaner with single methods per validation type for both scenario and node like `validateIdIsNotEmpty(implicit nodeId: String, isFragment: Option[Boolean])` and there we could differentiate between node/fragment/scenario based on the 3 isFragment states?

Reply:
IMO it would be nice to have general purpose id validation mechanism to avoid duplication of this case of `validateIdIsNotEmpty andThen validateIdIsNotBlank ...`. It should be double if we figure out how to handle this error creation strategies depended on context. Maybe we should have generic errors e.g. `EmptyId` / `BlankId` and each of them should have something like `validatedIdName` and then map errors to '$validatedIdName is mandatory and can't be empty`. WDYT?

Reply:
Ok, I see now. I've changed it to a consistent validation strategy with mapping to specific errors internally in the IdValidator.

Reply:
It reduced the part of code that was the most important for me, so it is almost perfect. IMO we could go one step further and remove even these error mapping part. We could achieve that by like I wrote extracting `validatedIdName` (which could be e.g. `Fragment name` inside `EmptyIdError` etc. and use it in a message. But this code duplication isn't crucial so I'm not forcing you to do that - your choice.